Make Google Cloud STT/TTS timeout configurable

eslavnov opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
Hello, this is my first ever PR for Home Assistant - I sure hope I've done everything correctly, haha!

The existing implementation has a hard-coded 10s timeout when waiting for the generated speech from Google Cloud TTS.
Because of that, the integration fails (error "400 Request contains an invalid argument") when you try to generate outputs larger than roughly 600-800 characters, since it would take longer than that for Google Cloud TTS to return a response. I've been hitting this issue constantly when trying to announce long responses coming from my LLM, for example.

The proposed change makes TTS Timeout configurable in the UI. The default value is set at 10 seconds (same as the existing implementation), but now users can increase it if needed to generate longer output. The tests have been updated, too.
<img width="238" alt="2025-01-26 12_14_39-Settings – Home Assistant" src="https://github.com/user-attachments/assets/f3a6a36c-2704-4eee-bbca-079f41f5fdfe" />

## Type of change
- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue:
- This PR is related to issue: Might be related to this https://github.com/home-assistant/core/issues/116983 (same error), but it's hard to say for sure since it was closed with little details.
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/37122

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`. (no changes)
- [X] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description. (no changes)

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

eslavnov committed:
Make TTS timeout configurable

eslavnov committed:
Fix strings

eslavnov committed:
Update tests

eslavnov committed:
Merge branch 'home-assistant:dev' into google_cloud_tts_timeout

frenck commented:
Honestly, I don't think this is an option that should work this way. It means you have to take into account what you are sending as a user (e.g., output of a template, could be unknown up front).

Instead, we could maybe consider a different default, error handling or logic to automatically set a sensible timeout depending on requested content?

I don't feel this is a burden we should put on an end-user.

eslavnov commented:
Thanks @frenck for the feedback! I agree with your description of the endstate, this would be great indeed.

However, I don't think that having this as an optional configurable parameter creates any extra burden for the users compared to the current state: their experience does not change except for allowing power users to change the timeout if it fails on their use cases. 

So I see this PR as an incremental improvement and not the end solution. But maybe it's a step too small and somebody needs to work on the proper endstate solution instead.

eslavnov commented:
Something to consider is how do we define the timeout based on the input? The response time will depend on the selected model, language and probably text, too. If we say "for each X chars add Y ms to the timeout" and undershoot it, we end up with the same issue for the end-user just with extra steps on our side. So we need to somehow estimate the response time with some generous extra buffer, while making sure we don't end up with unreasonably long timeouts...

tronikos commented on homeassistant/components/google_cloud/const.py:
```diff
@@ -19,7 +19,9 @@
 CONF_GAIN = "gain"
 CONF_PROFILES = "profiles"
 CONF_TEXT_TYPE = "text_type"
+CONF_TIMEOUT= "tts_timeout"
```
space before =

tronikos replied:
Remove tts_ prefix and use it in stt.py too.

tronikos commented on homeassistant/components/google_cloud/strings.json:
```diff
@@ -25,7 +25,8 @@
           "text_type": "Default text type",
-          "stt_model": "STT model"
+          "stt_model": "STT model",
+          "tts_timeout": "TTS timeout, seconds",
```
Remove `tts_` and `TTS ` prefixes and use it in stt.py too.

tronikos replied:
For consistency with other options, e.g. gain above, change to `timeout (in seconds)`

eslavnov committed:
Implement feedback

eslavnov committed:
Merge branch 'google_cloud_tts_timeout' of https://github.com/eslavnov/core into google_cloud_tts_timeout

eslavnov committed:
Fix STT timeout

eslavnov commented:
Hey @tronikos, thank you for the feedback, and pardon the delay - I got a bit distracted with a side quest :)
I've implemented all your proposed changes and updated the docs PR too - let me know if anything is missing!

eslavnov committed:
Fix STT timeout

eslavnov commented:
Hey @tronikos, thank you for the feedback, and pardon the delay - I got a bit distracted with a side quest :)
I've implemented all your proposed changes and updated the docs PR too - let me know if anything is missing!

eslavnov committed:
fix whitespace

tronikos commented on homeassistant/components/google_cloud/tts.py:
```diff
@@ -215,8 +217,11 @@ async def _async_get_tts_audio(
 
-        response = await self._client.synthesize_speech(request, timeout=10)
-
+        timeout = options[CONF_TIMEOUT]
```
Replace it with `options.get(CONF_TIMEOUT, DEFAULT_TIMEOUT)` and inline it?

eslavnov replied:
Oops, that was dumb of me - fixed, thanks!

eslavnov committed:
Inline tts timeout

tronikos commented on homeassistant/components/google_cloud/tts.py:
```diff
@@ -215,8 +217,8 @@ async def _async_get_tts_audio(
-        response = await self._client.synthesize_speech(request, timeout=10)
-
+        response = await self._client.synthesize_speech(request, timeout=options.get(CONF_TIMEOUT, DEFAULT_TIMEOUT))
+        
```
Remove extra whitespace

eslavnov committed:
Remove whitespace

tronikos commented:
Can you run ruff?

eslavnov committed:
Run ruff check and format

eslavnov commented:
@tronikos, my apologies and thank you for your **_enormous_** patience, fixed now! (run both ruff check & format, so should be good).

frenck commented on homeassistant/components/google_cloud/helpers.py:
```diff
@@ -150,6 +152,10 @@ def tts_options_schema(
                     )
                 ),
             ),
+            vol.Optional(
+                CONF_TIMEOUT,
+                default=defaults.get(CONF_TIMEOUT, DEFAULT_TIMEOUT),
+            ): NumberSelector(NumberSelectorConfig(min=1, max=60, step=1)),
```
My original stance remains on this one. I don't think this belong in a user-facing configuration.

If you really disagree, then I'm happy to validate my concern within a core architectural meeting.

../Frenck

eslavnov replied:
Thanks, @frenck! I trust your knowledge and expertise more than mine, so please feel free to reject this PR. 

I really don't have a horse in this race, it was just something that really frustrated me as a user so I thought I'd share the fix. I still believe that an extra option with a default value does not create any extra burden on users, but I do agree that we need to draw the line somewhere, otherwise we end up with a billion of various settings. 

So I see your point and I trust that you know best where to draw this line and what should be configurable or not 👍

frenck replied:
> I really don't have a horse in this race, it was just something that really frustrated me as a user so I thought I'd share the fix.

I appreciate that, don't get me wrong. But I think we should be smarter about it, and not drop the issue with the end user. For example, extended timeouts or change the timeouts depending on requested text length.

> I still believe that an extra option with a default value does not create any extra burden on users

Sorry, I think differently. People run into something not working and have to figure out how this would and that this is an timeout issue (and next how to extend the timeout).

../Frenck

eslavnov replied:
Hey Frenck, thanks for the pushback. I took some time to think and reflect on your point and I agree: our users deserve better. How about the following approach instead?

1. We keep the default timeout at 10s, so we don't affect the experience of current users.
2. In my tests 10s timeout allows me to generate 600-800 chars. This depends on the network, language, model, etc., so let's say we set a limit at 500 chars just to be safe.
3. If the text is less than the limit, we use the 10s timeout for the TTS (nothing changes compared to the current state).
4. If it is over 500 chars, for each extra character we add 0.1ms to the timeout (or something like that, I'll have to run some tests to figure out a good value). We also set a reasonable max timeout limit (60 seconds?) and we apply the lower of the two.
5. For the time being, we leave the STT component unchanged. Given how fast their STT is, 10 seconds there is plenty. 

This way we keep don't introduce any new user-facing settings, keep the existing behavior unchanged for short texts and improve the experience for long ones (since currently they simply timeout). The only downside I see is that since there are so many factors potentially affecting the latency, for some weird setups our extra timeout calculation logic might not yield a large enough value. But these would probably be really rare edge cases, so not a big deal. And in any case it's an improvement compared to the current state. 

What do you think, @frenck and @tronikos? I am happy to rework the PR, but I'd love for all f us to agree on the approach first, please 🙏
